src/Raven.Server/Documents/Queries/LuceneIntegration/TermsMatchQuery.cs

```diff
@@ -19,115 +22,191 @@ public TermsMatchQuery(string field, List<string> matches)
+                result.AddDetail(new Lucene.Net.Search.Explanation(_queryWeight, "queryWeight"));
+                return result;
+            }
 
-        private sealed class RavenTermsFilteredTermEnum : FilteredTermEnum
+            public override void Normalize(float norm)
+            {
+                _queryWeight *= norm;
+            }
+
+            public override Scorer Scorer(IndexReader reader, bool scoreDocsInOrder, bool topScorer, IState state)
+            {
+                Similarity similarity = _parent.GetSimilarity(_searcher);
+                if(_parent.Matches.Count > 128)
+                    return new LazyInitTermMatchScorer(_parent, reader, state, similarity);
+
+                var scorers = new Scorer[_parent.Matches.Count];
+                int index = 0;
+                foreach (var match in _parent.Matches)
+                {
+                    var termDocs = reader.TermDocs(new Term(_parent.Field, match),state);
+                    var scorer = new TermScorer(this, termDocs,similarity, reader.Norms(_parent.Field, state));
+                    if (scorer.NextDoc(state) != DocIdSetIterator.NO_MORE_DOCS)
+                    {
+                        scorers[index++] = scorer;
+                    }
+                }
+                return new DisjunctionMaxScorer(1.0f, similarity, scorers, index);
+            }
+
+            public override float GetSumOfSquaredWeights()
+            {
+                return _queryWeight * _queryWeight;
+            }
+
+            public override Query Query { get; }
+            public override float Value { get; }
+        }
+        
+        private class LazyInitTermMatchScorer : Scorer
         {
-            private readonly TermsMatchQuery _termsMatchQuery;
+            private readonly TermsMatchQuery _parent;
             private readonly IndexReader _reader;
             private readonly IState _state;
-            private bool _endEnum;
-            private int _pos;
+            private readonly Similarity _similarity;
+            private EagerTermMatchScorer _inner;
 
-            public RavenTermsFilteredTermEnum(TermsMatchQuery termsMatchQuery, IndexReader reader, IState state)
+            internal LazyInitTermMatchScorer(TermsMatchQuery parent, IndexReader reader, IState state, Similarity similarity) : base(similarity)
             {
-                _termsMatchQuery = termsMatchQuery;
+                _parent = parent;
                 _reader = reader;
                 _state = state;
-                if (_termsMatchQuery.Matches.Count == 0)
-                {
-                    _endEnum = true;
-                    return;
-                }
-                MoveToCurrentTerm(state);
+                _similarity = similarity;
             }
 
-            private void MoveToCurrentTerm(IState state)
+            private EagerTermMatchScorer InitIfNeeded()
```

Comment:
Maybe add `[MethodImpl(MethodImplOptions.AggressiveInlining)]` ?

Reply:
Probably not, this is small enough that we should let the JIT decide.